pallets/msa/src/offchain_storage.rs

```diff
@@ -0,0 +1,464 @@
+use crate::{pallet::OffchainIndexEventCount, Config, Event, Pallet, PublicKeyToMsaId};
+pub use common_primitives::msa::MessageSourceId;
+/// Offchain Storage for MSA
+use common_primitives::offchain::{
+	self as offchain_common, get_msa_account_lock_name, get_msa_account_storage_key_name,
+	LockStatus, MSA_ACCOUNT_LOCK_TIMEOUT_EXPIRATION,
+};
+use frame_support::RuntimeDebugNoBound;
+use frame_system::pallet_prelude::BlockNumberFor;
+use parity_scale_codec::{Decode, Encode};
+use sp_core::serde::{Deserialize, Serialize};
+extern crate alloc;
+use alloc::string::String;
+use sp_io::offchain_index;
+use sp_runtime::{
+	offchain::{
+		storage::StorageValueRef,
+		storage_lock::{BlockAndTime, StorageLock, Time},
+		Duration,
+	},
+	traits::One,
+	Saturating,
+};
+use sp_std::{collections::btree_map::BTreeMap, fmt::Debug, vec, vec::Vec};
+
+/// Block event storage prefix
+const BLOCK_EVENT_KEY: &[u8] = b"frequency::block_event::msa::";
+
+/// Offchain index for MSA events count
+const BLOCK_EVENT_COUNT_KEY: &[u8] = b"frequency::block_event::msa::count::";
+
+/// Lock expiration timeout in in milli-seconds for initial data import msa pallet
+const MSA_INITIAL_LOCK_TIMEOUT_EXPIRATION: u64 = 3000;
+
+/// Lock expiration block for initial data import msa pallet
+const MSA_INITIAL_LOCK_BLOCK_EXPIRATION: u32 = 30;
+
+/// Lock name for initial data index for msa pallet
+const MSA_INITIAL_LOCK_NAME: &[u8; 28] = b"Msa::ofw::initial-index-lock";
+
+/// storage name for initial data import storage
+pub const MSA_INITIAL_INDEXED_STORAGE_NAME: &[u8; 25] = b"Msa::ofw::initial-indexed";
+
+/// Lock name for last processed block number events
+const LAST_PROCESSED_BLOCK_LOCK_NAME: &[u8; 35] = b"Msa::ofw::last-processed-block-lock";
+
+/// lst processed block storage name
+pub const LAST_PROCESSED_BLOCK_STORAGE_NAME: &[u8; 30] = b"Msa::ofw::last-processed-block";
+
+/// Lock expiration timeout in in milli-seconds for last processed block
+const LAST_PROCESSED_BLOCK_LOCK_TIMEOUT_EXPIRATION: u64 = 5000;
+
+/// Lock expiration for last processed block
+const LAST_PROCESSED_BLOCK_LOCK_BLOCK_EXPIRATION: u32 = 20;
+
+/// number of previous blocks to check to mitigate offchain worker skips processing any block
+const NUMBER_OF_PREVIOUS_BLOCKS_TO_CHECK: u32 = 5u32;
+
+/// number of blocks to explore when trying to find the block number from block hash
+const NUMBER_OF_BLOCKS_TO_EXPLORE: u32 = 1000;
+
+/// HTTP request deadline in milliseconds
+pub const HTTP_REQUEST_DEADLINE_MILLISECONDS: u64 = 2000;
+
+/// LOCAL RPC URL and port
+/// warning: this should be updated if rpc port is set to anything different from 9944
+pub const RPC_FINALIZED_BLOCK_REQUEST_URL: &'static str = "http://localhost:9944";
```

Comment:
It would have been great if we could pass this as an input param but not sure if it's possible but open to suggestions.

Reply:
We should see if we cannot derive it, while `9944` is the default, it uses the `rpc-port` cli arg.

Reply:
created another issue to track this https://github.com/LibertyDSNP/frequency/issues/1846